### PR TITLE
Use pyproject.toml to setup test discovery

### PR DESCRIPTION
**Pull request by theo-brown**
I've added lines in `pyproject.toml` to configure pytest for automatically discovering tests, following #441.

This should:
- Ignore `torax/tests/scripts`
- Ignore `torax/tests/test_data`
- Ignore any directory with `lib` in the name
- Collect all files with a `.py` extension from directories called `tests`, at any level

This necessitated changing the names of the `runtime_params` test files in `transport_models` and `sources` to prevent filename conflicts.

I've modified the Github action to run the tests via auto-discovery.

I have deleted duplicate `conftest.py` files, and consolidated them into on file in the `torax/` directory. I believe this will be shared between all the necessary tests, but I'm not completely sure.

### CODE REVIEW

**Comment by jcitrin** [comment]
Thanks, this is very helpful!

Am testing it and realized that there are some Equinox issues, and also at least one unused broken test that we need to clean up. Possibly we'll want to add some more exclusions too. Will debug on our side and push to this PR.

**Comment by theo-brown** [comment]
Yep, I think there were a few tests that were not previously being picked up by CI, but now are, that are failing.

I didn't spot the Equinox issues my side, but I didn't start digging into making all the tests work, so it could just be that I didn't look hard enough.

Definitely scope for more inclusions/exclusions - I wasn't certain which of the tests were purposefully vs accidentally being skipped in CI, so I decided to just push this to a PR so that others could work on it to.

**Comment by jcitrin** [comment]
The Equinox issue was a non-issue, was just a non-updated venv on my side.

We'll take a look at further exclusions / cleanups. Thanks again!

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -62,40 +62,9 @@ jobs:
...
-          torax/tests/math_utils.py \
-          torax/tests/persistent_cache.py \
-          torax/tests/physics.py \
-          torax/tests/test_run_simulation_main.py \
-          torax/tests/sim_custom_sources.py \
-          torax/tests/sim_time_dependence.py \
-          torax/tests/sim.py \
-          torax/tests/state.py \
-          torax/transport_model/tests/qlknn_wrapper.py \
-          torax/transport_model/tests/transport_model.py \
```
[the remark this answers is not in the document]
Comment (jcitrin) [reply .github/workflows/pytest.yml:98]: Thanks for flagging. It won't but we're working on it.

[the remark this answers is not in the document]
Comment (theo-brown) [reply .github/workflows/pytest.yml:98]: (relevant commit 029c24c0c26332f70838c7ad3ad73ebdca8f4849)

**Comment by theo-brown** [comment]
I ran `pytest` locally and all the tests that were run/discovered passed except from the following.

https://github.com/google-deepmind/torax/blob/7f675ab8f207c28ada41737bf82aaebc81ded064/torax/tests/test_run_simulation_main.py#L95C3-L116C76

**Comment by Nush395** [comment]
> I ran `pytest` locally and all the tests that were run/discovered passed except from the following (fails on the last assertion in the highlighted block).
> 
> https://github.com/google-deepmind/torax/blob/7f675ab8f207c28ada41737bf82aaebc81ded064/torax/tests/test_run_simulation_main.py#L95C3-L116C76

Thanks Theo! Will look into what the issue is there.

**Comment by Nush395** [comment]
> I ran `pytest` locally and all the tests that were run/discovered passed except from the following (fails on the last assertion in the highlighted block).
> 
> https://github.com/google-deepmind/torax/blob/7f675ab8f207c28ada41737bf82aaebc81ded064/torax/tests/test_run_simulation_main.py#L95C3-L116C76

Have found the issue here and it's quite fiddly, thanks for uncovering! That assertion using the captured output only works when `test_main_app_runs` is run in isolation and doesn't work when run with the other tests in the module. This hasn't been an issue in the workflow setup because of the `-n auto` flag which spins up num_cpus to run the test with and it seems to be the case that it works there as the different tests in this module end up on different processes and don't interact. 

There are two fixes we can try:
1. Add `-n=auto` to the `pytest.ini_options` to this CL.
2. I will do a separate change that will modify this check that this string is present in captured output to something more robust.

For the purposes of submitting and testing this CL we could go with 1 to unblock you and allow you to test and submit and I will work on 2 in the background later today. WDYT?

**Comment by theo-brown** [comment]
Yep, I think this has fixed it, thanks @Nush395.
However, I think the addition of the QuaLiKiz model (#473) might necessitate some changes. In particular, I think there might need to be an import check or a flag for whether qualikiz_tools is installed? This might just be an issue with running them via pytest vs absl.

https://github.com/google-deepmind/torax/blob/a23ec9b5f4f8dab98e7d525f6c5c622f7d6b8eb5/torax/transport_model/tests/qualikiz_wrapper.py#L17

**Comment by jcitrin** [comment]
That check is done here: https://github.com/google-deepmind/torax/blob/a23ec9b5f4f8dab98e7d525f6c5c622f7d6b8eb5/torax/config/build_sim.py#L44

**Comment by jcitrin** [comment]
Ah, that's for the test, which is indeed different. Will look into it

**Comment by jcitrin** [comment]
Fixed by #481 . Can sync and try again

**Comment by theo-brown** [comment]
Yep, all good. 488 passed, 9 skipped, 6 warnings.

**Comment by Nush395** [comment]
Thanks for the change Theo, FYI I had to add

```
"--ignore=torax/transport_model/qualikiz_wrapper.py",
```
to the ini_options to get it to work as well.

**Comment by theo-brown** [comment]
Interesting - that looks like it's a different/parallel fix to #481?

**Pull request merged**